src/m365/spo/commands/file/file-sharinglink-add.ts

```diff
@@ -0,0 +1,167 @@
+import { Logger } from '../../../../cli/Logger';
+import GlobalOptions from '../../../../GlobalOptions';
+import request, { CliRequestOptions } from '../../../../request';
+import { formatting } from '../../../../utils/formatting';
+import { validation } from '../../../../utils/validation';
+import SpoCommand from '../../../base/SpoCommand';
+import commands from '../../commands';
+import { GraphFileDetails } from './GraphFileDetails';
+
+interface CommandArgs {
+  options: Options;
+}
+
+interface Options extends GlobalOptions {
+  webUrl: string;
+  fileId?: string;
+  fileUrl?: string;
+  type: string;
+  expirationDateTime: string;
+  scope?: string;
+}
+
+class SpoFileSharingLinkAddCommand extends SpoCommand {
+  private static readonly types: string[] = ['view', 'edit'];
+  private static readonly scopes: string[] = ['anonymous', 'organization'];
+
+  public get name(): string {
+    return commands.FILE_SHARINGLINK_ADD;
+  }
+
+  public get description(): string {
+    return 'Creates a new sharing link for a file';
+  }
+
+  constructor() {
+    super();
+
+    this.#initTelemetry();
+    this.#initOptions();
+    this.#initValidators();
+    this.#initOptionSets();
+  }
+
+  #initTelemetry(): void {
+    this.telemetry.push((args: CommandArgs) => {
+      Object.assign(this.telemetryProperties, {
+        fileId: typeof args.options.fileId !== 'undefined',
+        fileUrl: typeof args.options.fileUrl !== 'undefined',
+        expirationDateTime: typeof args.options.expirationDateTime !== 'undefined',
+        scope: typeof args.options.scope !== 'undefined'
+      });
+    });
+  }
+
+  #initOptions(): void {
+    this.options.unshift(
+      {
+        option: '-u, --webUrl <webUrl>'
+      },
+      {
+        option: '-i, --fileId [fileId]'
+      },
+      {
+        option: '-f, --fileUrl [fileUrl]'
+      },
+      {
+        option: '--type <type>',
+        autocomplete: SpoFileSharingLinkAddCommand.types
+      },
+      {
+        option: '--expirationDateTime [expirationDateTime]'
+      },
+      {
+        option: '--scope [scope]',
+        autocomplete: SpoFileSharingLinkAddCommand.scopes
+      }
+    );
+  }
+
+  #initValidators(): void {
+    this.validators.push(
+      async (args: CommandArgs) => {
+        const isValidSharePointUrl: boolean | string = validation.isValidSharePointUrl(args.options.webUrl);
+        if (isValidSharePointUrl !== true) {
+          return isValidSharePointUrl;
+        }
+
+        if (args.options.fileId && !validation.isValidGuid(args.options.fileId)) {
+          return `${args.options.fileId} is not a valid GUID`;
+        }
+
+        if (args.options.type &&
```

Comment:
Since type is a required option, we don't have to check first if it has a value.